gravitee-apim-rest-api/gravitee-apim-rest-api-service/src/main/java/io/gravitee/apim/infra/membership/PrimaryOwnerDomainServiceImpl.java

```diff
@@ -0,0 +1,141 @@
+import io.gravitee.rest.api.service.exceptions.PrimaryOwnerNotFoundException;
+import io.gravitee.rest.api.service.exceptions.TechnicalManagementException;
+import java.util.Optional;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Service;
+
+@Slf4j
+@Service
+public class PrimaryOwnerDomainServiceImpl implements ApiPrimaryOwnerDomainService {
+
+    private final RoleRepository roleRepository;
+    private final MembershipRepository membershipRepository;
+    private final UserCrudService userCrudService;
+    private final GroupRepository groupRepository;
+
+    public PrimaryOwnerDomainServiceImpl(
+        @Lazy RoleRepository roleRepository,
+        @Lazy MembershipRepository membershipRepository,
+        UserCrudService userCrudService,
+        @Lazy GroupRepository groupRepository
+    ) {
+        this.roleRepository = roleRepository;
+        this.membershipRepository = membershipRepository;
+        this.userCrudService = userCrudService;
+        this.groupRepository = groupRepository;
+    }
+
+    @Override
+    public PrimaryOwnerEntity findApiPrimaryOwner(ExecutionContext executionContext, String apiId) {
+        return findPrimaryOwnerRole(executionContext.getOrganizationId())
+            .flatMap(role ->
+                findPrimaryOwnerMembership(apiId, role)
+                    .flatMap(membership ->
+                        switch (membership.getMemberType()) {
+                            case USER -> buildUserPrimaryOwner(membership);
+                            case GROUP -> buildGroupPrimaryOwner(membership, role.getId());
+                        }
+                    )
+            )
+            .orElseThrow(() -> new PrimaryOwnerNotFoundException(apiId));
+    }
+
+    private Optional<Role> findPrimaryOwnerRole(String organizationId) {
+        try {
+            return roleRepository.findByScopeAndNameAndReferenceIdAndReferenceType(
+                RoleScope.API,
+                SystemRole.PRIMARY_OWNER.name(),
+                organizationId,
+                RoleReferenceType.ORGANIZATION
+            );
+        } catch (TechnicalException e) {
+            log.error("An error occurs while trying to get primary owner role for [organizationId={}]", organizationId, e);
+            throw new TechnicalManagementException(e);
+        }
+    }
+
+    private Optional<Membership> findPrimaryOwnerMembership(String apiId, Role role) {
+        try {
+            return membershipRepository.findByReferenceAndRoleId(MembershipReferenceType.API, apiId, role.getId()).stream().findFirst();
+        } catch (TechnicalException e) {
+            log.error("An error occurs while trying to get primary owner for [apiId={}]", apiId, e);
+            throw new TechnicalManagementException(e);
+        }
+    }
+
+    private Optional<PrimaryOwnerEntity> buildUserPrimaryOwner(Membership membership) {
```

Comment:
with `build` it's not obvious we are in fact fetching something.
I would prefer something like
```suggestion
    private Optional<PrimaryOwnerEntity> findPrimaryOwnerFrom(Membership membership) {
```